### PR TITLE
[User Settings] Implement new Profile view

### LINKED ISSUE #1492: [User Settings] Implement new Profile view
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/master/CONTRIBUTING.md) for onboarding!

___

We're adding a new profile view that will display a right-docked modal with a user's avatar, primary login, as well as their first name, last name, pronouns, and timezone (as applicable, depending on what data the user has filled out. This page will mimic the current [SettingsPage.js](https://github.com/Expensify/Expensify.cash/blob/master/src/pages/SettingsPage.js) view with a few key differences
- No `Settings` title 
- No `Sign Out` button 
- New (non-editable) fields with titles that are displayed/hidden depending on what the user has filled out in the `personalDetails` onyx key) 
     - Email
     - Phone Number
     - Preferred Pronouns 
     - Local Time

**Deliverables:**
- Update HeaderView.js to display a clickable icon associated with the given chat (`report.icons`) 
   - 
   - You can reuse the [MultipleAvatars.js](https://github.com/Expensify/Expensify.cash/blob/master/src/components/MultipleAvatars.js#L19) component for this 
   - If the chat has multiple participants (`report.participants > 1`) clicking on the icon should do nothing 
   - Otherwise, the chat is a direct message (having only one participant) and clicking on the icon should open the `Profile` page for that `report.participant`
       - This works similarly to how the Settings page opens up when a user clicks on their own avatar in the sidebar
- Add two new routes to [Routes.js](https://github.com/Expensify/Expensify.cash/blob/master/src/ROUTES.js)
    - `Profile: '/profile/:login`
    - ``getProfileRoute: login => `/profile/${login}` ``
- Create the right-docked modal "Profile" page following the example of the "Settings" page and refactoring as necessary 
    -  The profile page should take in the `login` associated with the `report.participant` as a prop
    - That login will be used to get the associated `personalDetail` object for that user from the `personalDetails` onyx key 
        - That object will contain the information to be displayed for this user on their Profile page.  

### Mobile 
| No Details Filled Out  | All Details Filled Out|
| ------------- | ------------- |
|  |  |

### Web/Desktop
| No Details Filled Out  | All Details Filled Out|
| ------------- | ------------- |
| image | image |

### Upwork Link: https://www.upwork.com/jobs/~01f46c14a330e02af6

### PR DESCRIPTION
**Pull request by Maftalion**
<If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit.>

### Details
- Set up new Profile route, uses the personalDetails tied to the login or the report (as a fallback)
- Redo MultipleAvatars to be able to be wrapped without getting messed up

### Fixed Issues
<Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing>
Fixes https://github.com/Expensify/Expensify.cash/issues/1492

### Tests
1. start a chat with a verified user and click their avatar in the header to get routed to see their profile info
2. start a chat with an unverified user and click their avatar in the header to get routed to see their profile info and minimally see their avatar & email/login
3. start a chat with multiple users and confirm that clicking their avatar does nothing

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Video Walkthrough

https://user-images.githubusercontent.com/12075600/109588212-3dd97080-7abd-11eb-8f6b-bff8ce58152f.mov

#### Web

Screen Shot 2021-03-01 at 6 29 29 PM

#### Mobile Web

Screen Shot 2021-03-01 at 6 41 46 PM

#### Desktop

Screen Shot 2021-03-01 at 6 27 44 PM

#### iOS

Screen Shot 2021-03-01 at 6 24 07 PM

#### Android

Screen Shot 2021-03-01 at 6 26 40 PM

### CODE REVIEW

File: src/pages/ProfilePage.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    // timezone of the current user from their personal details
+    timezone: PropTypes.string,
+});
+
+const reportType = PropTypes.shape({
+    reportID: PropTypes.number,
+    reportName: PropTypes.string,
+    participants: PropTypes.arrayOf(PropTypes.string),
+    icons: PropTypes.arrayOf(PropTypes.string),
+});
```
Comment (Maftalion) [inline src/pages/ProfilePage.js:41]: I suggest create a file/folder to contain all these types based off onyx keys since they will be re-used a lot across components.

File: src/components/MultipleAvatars.js [hunk 2]
```diff
@@ -16,62 +17,59 @@ const defaultProps = {
     optionIsFocused: false,
 };
 
-const MultipleAvatars = props => (
-    <>
-        {props.avatarImageURLs !== undefined && (
-            <>
+const MultipleAvatars = ({avatarImageURLs, optionIsFocused}) => {
```
Comment (Maftalion) [inline src/components/MultipleAvatars.js:20]: Had to redo this whole component, it was causing issues when wrapped by the Pressable. Now also defaults to the regular Avatar if only 1 is passed.
SS of issue below:
Screen Shot 2021-03-01 at 6 52 26 PM

**Comment by shawnborton** [comment]
Do we want the rightDocked profile view to have a modal title? Something like "Details" or "Profile"?

**Review by pecanoro — DISMISSED** [review]
@marcaaron @roryabraham  can any of you take a look at this? React is not really my strength so I am not sure if I am missing something.

File: src/components/MultipleAvatars.js [hunk 3]
```diff
@@ -16,62 +17,59 @@ const defaultProps = {
...
+    if (!avatarImageURLs?.length) {
```
Comment (marcaaron) [inline src/components/MultipleAvatars.js:21]: Just a heads up, we are not using any optional chaining as per our style guide so it would be preferable to avoid it if we can.

File: src/pages/ProfilePage.js (hunk 1 above)
Comment (marcaaron) [reply src/pages/ProfilePage.js:41]: Nice callout. I think they probably already are used in a few places and we could do this in a clean up / refactoring issue. Probably doesn't need to be done right now.

File: src/pages/ProfilePage.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+                onCloseButtonPress={redirectToLastReport}
+            />
+            <View
+                pointerEvents="box-none"
+                style={[
+                    styles.profilePageContainer,
+                ]}
+            >
+                <View>
+                    <View style={styles.settingsWrapper}>
```
Comment (marcaaron) [inline src/pages/ProfilePage.js:72]: This name should be generalized if we are reusing styles from the "settings" page.

File: src/pages/ProfilePage.js [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                        <View
+                            style={[styles.avatarLarge, styles.mb3]}
+                        >
+                            <Avatar
+                                style={[styles.avatarLarge]}
+                                source={profileDetails.avatarURL}
+                            />
+                        </View>
+                        <Text style={[styles.settingsDisplayName, styles.mt1, styles.mb6]} numberOfLines={1}>
```
Comment (marcaaron) [inline src/pages/ProfilePage.js:81]: Same here.

File: src/pages/ProfilePage.js [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+                            </Text>
+                        </View>
+                        )}
+                    </View>
+                </View>
+
+                <Text style={[styles.profilePageSectionVersion]} numberOfLines={1}>
+                    v
+                    {version}
+                </Text>
```
Comment (marcaaron) [inline src/pages/ProfilePage.js:124]: I don't think we want the app version to exist on this screen.

File: src/pages/ProfilePage.js [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        personalDetails: {
+            key: ONYXKEYS.PERSONAL_DETAILS,
+        },
+    }),
+    withOnyx({
+        currentlyViewedReportID: {
+            key: ONYXKEYS.CURRENTLY_VIEWED_REPORTID,
+        },
+    }),
+    withOnyx({
```
Comment (marcaaron) [inline src/pages/ProfilePage.js:145]: `withOnyx()` can subscribe to multiple keys at once so we probably do not need 3 of these.
For the `currentlyViewedReportID` if we pass it as a match param instead then we can use it to fetch the report.

However, again, I am confused about why we are basing this on the report participants. It doesn't seem necessary to look at the report at all. If we are routing to `/profile/:login` then we need only look at the `:login` match param and then locate it in the `personalDetails`. 

IMO this solution is more complex since instead of getting the details by doing `this.props.personalDetails[this.props.match.params.login]` we are subscribing to three separate Onyx keys.

Please let me know if I'm missing something...

**Comment by NikkiWines** [comment]
@shawnborton not opposed to adding a title, but just a note that we're already planning on using the `Profile` title for the Settings > Profile view, so it might be confusing to use that here. What if it was something like `<Display Name>'s Details` or even `<Display Name>'s Profile` to ensure there's a distinction? 

image

**Review by marcaaron — CHANGES_REQUESTED** [review]
Thanks for your work here. To be honest, I much preferred the original proposal. This implementation does not seem flexible enough and will likely need to be refactored eventually.

File: src/pages/ProfilePage.js (hunk 6 above)
Comment (NikkiWines) [reply src/pages/ProfilePage.js:124]: Ah, the issue mockups do show the app version on the screen, so I think this is fine actually. Though I agree it might be a little out of place to have them here.

File: src/ROUTES.js [hunk 8]
```diff
@@ -13,4 +13,5 @@ export default {
     SET_PASSWORD: '/setpassword/:validateCode',
     SIGNIN: '/signin',
     NOT_FOUND: '/404',
+    PROFILE: '/profile',
```
[the remark this answers is not in the document]
Comment (NikkiWines) [reply src/ROUTES.js:16]: @Maftalion I think Marc brings up some good points here - I hadn't considered the implications of deep-linking when I gave you the go-ahead to use the simplified route that you proposed. 

It does seem like the `profile/:login` route is the better solution in this case.

**Comment by Maftalion** [comment]
Sounds good @NikkiWines @marcaaron. I can switch to back to the original implementation but here is some background info on what led to this implementation:

1. RightDockedModals don't current support dynamic routes.
Since the visibility value is set to `isVisible={currentURL === route}`, `/profile/:login` would never equal to the actual route like `/profile/aftalion.matt@gmail.com`.
With the original implementation I solved this by adding an optional `isDynamicRoute|boolean` prop that would split out the route from the `:` to compare `/profile/` to `/profile/` but not sure how safe this approach is against subroutes so would need to look into that as well.

2. Fallback for user not existing in `PERSONAL_DETAILS`.
The idea was that it would use the login/avatar from the report as a fallback which led to the option of login not being passed at all since all that data can be pulled from the report as well, but as you mentioned it would rely heavily on `currentlyViewedReportID`.

So to recap what I would be doing based on the feedback would be:
1. Re-implementing the dynamic routes on rightdockedmodals in order for `/profile/:login` to work.
2. Setting up the route to be `/profile/:login` and using `match.params` to get the user details from `PERSONAL_DETAILS`.

**Questions**
1. What do we do for users that don't show up in `PERSONAL_DETAILS` (or in prod would they always exist?)
     basically if I just click to start a new chat with a random unverified user, they won't show up in the`PERSONAL_DETAILS` 
which is why this implementation was set up to pull participant/avatar from the report as a fallback.

File: src/pages/ProfilePage.js (hunk 7 above)
Comment (Maftalion) [reply src/pages/ProfilePage.js:145]: interesting, I tried adding all the keys into one and it was complaining so I kept them separated but maybe I did something wrong. but as you mentioned, with the original implementation there would only need to be one key

**Comment by shawnborton** [comment]
I think just the title of "Details" would be good then.

**Comment by marcaaron** [comment]
>RightDockedModals don't current support dynamic routes.
Since the visibility value is set to isVisible={currentURL === route}, /profile/:login would never equal to the actual route like /profile/aftalion.matt@gmail.com.

I would suggest we solve this by just checking to see if the `currentURL.includes('/profile')` I think this would then tell us if this is the correct route without looking for an exact match.

>With the original implementation I solved this by adding an optional isDynamicRoute|boolean prop that would split out the route from the : to compare /profile/ to /profile/ but not sure how safe this approach is against subroutes so would need to look into that as well.

I think `isDynamicRoute` is not needed. We can just add a `withRouter()` inside the profile page and access whatever variables are on the route via `match.params`.

>Fallback for user not existing in PERSONAL_DETAILS. The idea was that it would use the login/avatar from the report as a fallback which led to the option of login not being passed at all since all that data can be pulled from the report as well, but as you mentioned it would rely heavily on currentlyViewedReportID.

This is an understandable instinct. But why would the user be missing from `personalDetails`? There must be some reason they are not there when we expect them to be? And so, I think it would be better to ensure they are there and not fallback to the report. I believe this was an oversight when building out the "invite a friend" feature and we can do better than a work-around.

>What do we do for users that don't show up in PERSONAL_DETAILS (or in prod would they always exist?) basically if I just click to start a new chat with a random unverified user, they won't show up in thePERSONAL_DETAILS which is why this implementation was set up to pull participant/avatar from the report as a fallback. Possible solution would be to just use the login to display the email/phone and have a default avatar displayed?

Taking a quick look at the existing code it seems like they are added when processing the `chatList` here:
https://github.com/Expensify/Expensify.cash/blob/701821e1db701316da746de6bfb2287e1392602e/src/libs/actions/Report.js#L196-L197

And also indirectly when we create a new chat here:
https://github.com/Expensify/Expensify.cash/blob/701821e1db701316da746de6bfb2287e1392602e/src/libs/actions/Report.js#L456-L470

Since this code should run and get the personal details:
https://github.com/Expensify/Expensify.cash/blob/701821e1db701316da746de6bfb2287e1392602e/src/libs/actions/Report.js#L711-L715

But when we get here the personal details are not found for this user
https://github.com/Expensify/Expensify.cash/blob/701821e1db701316da746de6bfb2287e1392602e/src/libs/actions/PersonalDetails.js#L151-L155

Without going into too much detail about why that's the case (probably the details just don't exist yet) this seems like a good place to provide some fallback details for a user when the API doesn't find any and I think will solve our issue.

**Comment by marcaaron** [comment]
One idea for how to do the fallback would be look at each email to see if the details were returned from the response and then just add fallback ones with a login only before calling `formatPersonalDetails()` [here](https://github.com/Expensify/Expensify.cash/blob/701821e1db701316da746de6bfb2287e1392602e/src/libs/actions/PersonalDetails.js#L154)

```js
if (!data[participantEmail]) {
    details[participantEmail] = {login: participantEmail};
}
```

**Comment by Maftalion** [comment]
@marcaaron Thanks for your detailed response. I ran into the login key not existing in PERSONAL_DETAILS which led to that workaround but if we can assume it would always exist then it definitely cleans things up and decouples the profile page from reports.

>I would suggest we solve this by just checking to see if the currentURL.includes('/profile') I think this would then tell us if this is the correct route without looking for an exact match.

Not sure what you mean by this based on the way the modals are currently set up:
https://github.com/Expensify/Expensify.cash/blob/aa11bfa52a47799f6fb7950e41b4e392a0d8a748/src/pages/home/HomePage.js#L115-L117
https://github.com/Expensify/Expensify.cash/blob/aa11bfa52a47799f6fb7950e41b4e392a0d8a748/src/components/RightDockedModal.js#L30-L41

The only values available are the route `/profile/:login` and the currentURL ex. `/profile/aftalion.matt@gmail.com`. But i think based on your suggestion I can connect the router and use `match.path` instead of currentURL which would handle dynamic routes.

> One idea for how to do the fallback would be look at each email to see if the details were returned from the response and then just add fallback ones with a login only before calling formatPersonalDetails() here

Regarding fallback, is that something you want me to add in or out of scope?

**Comment by Maftalion** [comment]
@marcaaron @NikkiWines updated based on feedback.

@shawnborton here's a screenshot with 'Details' as the title:
Screen Shot 2021-03-02 at 3 30 48 PM

**Comment by marcaaron** [comment]
>based on your suggestion I can connect the router and use match.path instead of currentURL which would handle dynamic routes

We can leave everything the same but pass `/profile` for the `route` prop instead of `/profile/:login` to `RightDockedModal`. Then change this to a fuzzy match (so ignoring extra params) instead of exact match...

https://github.com/Expensify/Expensify.cash/blob/288402e477e4ed6c9fd79c9ab0816207ac21d737/src/components/RightDockedModal.js#L30-L36

I think there's no need to look at `path` in `RightDockedModal`. We should do this in the view it will wrap instead since that's where we will handle the route param 🤔 

>Regarding fallback, is that something you want me to add in or out of scope?

Would be great to add in now if it's not too much trouble.

**Comment by NikkiWines** [comment]
@shawnborton I think it might be nice to have a bit more distinction (with the font weight) between the header `Email`/`Pronouns`/etc and the actual content, what do you think? 

image

**Review by NikkiWines — CHANGES_REQUESTED** [review]
Looking good, one minor comment

File: src/pages/ProfilePage.js [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+                                />
+                            </View>
+                            <Text style={[styles.displayName, styles.mt1, styles.mb6]} numberOfLines={1}>
+                                {profileDetails.displayName
+                                    ? profileDetails.displayName
+                                    : null}
+                            </Text>
+                            {profileDetails.login && (
+                            <View style={[styles.mb6, styles.profilePageSectionContainer]}>
+                                <Text style={[styles.profilePageLabel, styles.mb2]} numberOfLines={1}>
```
Comment (shawnborton) [inline src/pages/ProfilePage.js:87]: We should use the style `formLabel` for this, this way we standardize on the same field heading everywhere. cc @NikkiWines - this will make the text look bold.

File: src/pages/ProfilePage.js [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+                                    {Str.isSMSLogin(profileDetails.login) ? 'Phone Number' : 'Email'}
+                                </Text>
+                                <Text style={[styles.profilePageLabel]} numberOfLines={1}>
```
Comment (shawnborton) [inline src/pages/ProfilePage.js:90]: For this, I think we can just use `styles.textP`

File: src/pages/ProfilePage.js (hunk 6 above)
Comment (shawnborton) [reply src/pages/ProfilePage.js:124]: I agree, we only want the app version displayed on your own personal Settings page.

File: src/styles/styles.js [hunk 11]
```diff
@@ -768,33 +768,32 @@ const styles = {
         borderRightWidth: 0,
     },
 
+    singleAvatarContainer: {
+        marginTop: -2,
```
Comment (shawnborton) [inline src/styles/styles.js:772]: Can you explain what exactly this does?

Comment (Maftalion) [reply src/styles/styles.js:772]: the previous implementation had something similar: https://github.com/Expensify/Expensify.cash/pull/1593/files#diff-8ab8ba45ab927f1430b8d1209d187e003a9cdfe476f809e4483ef64cdba253ebL786

It basically just nudges it up a bit, when I re-did the component I tried to get it to match exactly what was there before. Here's a SS with & without this style (lemme know if you prefer to remove entirely):
Screen Shot 2021-03-03 at 12 05 26 AM

**Comment by Maftalion** [comment]
@shawnborton updated SS with new styles:
Screen Shot 2021-03-03 at 12 02 41 AM

**Comment by shawnborton** [comment]
Looks great, thanks! 

One thing I am noticing too is that it would be nice if the "Details" header was slightly larger, perhaps the same font size that we use for "Chats" on the home screen.

File: src/styles/styles.js (hunk 11 above)
Comment (shawnborton) [reply src/styles/styles.js:772]: Got it - I think it's fine without nudging it up. As long as everything is vertically centered in that area, we should be good.

File: src/components/MultipleAvatars.js [hunk 12]
```diff
@@ -1,7 +1,9 @@
 import React, {memo} from 'react';
 import PropTypes from 'prop-types';
 import {Image, Text, View} from 'react-native';
+import _ from 'lodash';
```
Comment (marcaaron) [inline src/components/MultipleAvatars.js:4]: We usually do `import lodashGet from 'lodash.get';` so it would be good to keep that pattern for consistency.

File: src/components/MultipleAvatars.js [hunk 13]
```diff
@@ -16,62 +18,59 @@ const defaultProps = {
     optionIsFocused: false,
 };
 
-const MultipleAvatars = props => (
-    <>
-        {props.avatarImageURLs !== undefined && (
-            <>
+const MultipleAvatars = ({avatarImageURLs, optionIsFocused}) => {
+    if (!_.get(avatarImageURLs, 'length')) {
```
Comment (marcaaron) [inline src/components/MultipleAvatars.js:22]: We tend to discourage using `lodashGet()` to access stuff like array properties and it is more common to see a default to an empty array first and then

```js
if (!avatarImagesURLs.length) {
```

**Review by marcaaron — COMMENTED** [review]
Just have a couple of small comments. This looks great to me in my testing so far.

File: src/components/MultipleAvatars.js (hunk 13 above)
Comment (Maftalion) [reply src/components/MultipleAvatars.js:22]: sounds good, I'll remove the get call. don't remember exactly why I added this check but maybe late-night-me just wanted a sanity check to not crash if a primitive type was somehow passed in as the prop

**Comment by Maftalion** [comment]
👍  Updated. Lemme know how I should proceed with the header, it's using the same component that shows up on all the other screens (other than Chats on the homescreen)

File: src/pages/ProfilePage.js (hunk 6 above)
Comment (NikkiWines) [reply src/pages/ProfilePage.js:124]: @Maftalion - let's remove the app version from this page.

**Comment by Maftalion** [comment]
@NikkiWines @shawnborton removed the version from the profile page

**Review by marcaaron — DISMISSED** [review]

**Comment by NikkiWines** [comment]
@Maftalion just a note that we just merged https://github.com/Expensify/Expensify.cash/pull/1573 which changes timezone from a string to an object, so it might be wise to update `src/pages/ProfilePage.js` to account for that change (even though for the moment timeZone only exists in `MY_PERSONAL_DETAILS` and not `PERSONAL_DETAILS` (I'm working on that though))

**Comment by Maftalion** [comment]
@NikkiWines updated to work with the new `PERSONAL_DETAILS` changes

**Review by NikkiWines — APPROVED** [review]
Tested with fully set personal details and it looks good!

image
image

**Review by pecanoro — APPROVED** [review]

**Review by marcaaron — APPROVED** [review]

**Comment by shawnborton** [comment]
Sorry for not catching this earlier, but I think tapping both the avatar **and** the name in the header should show the profile view. Currently you can only tap the avatar to access it. 

Would this be hard to implement as a follow up?
